cmd/composer-cli/blueprints/save.go

```diff
@@ -59,78 +69,110 @@ func saveToml(cmd *cobra.Command, args []string) (rcErr error) {
+func saveCommit(cmd *cobra.Command, name, commit string) error {
+	if root.JSONOutput {
+		_, resp, err := root.Client.GetBlueprintChangeJSON(name, commit)
+		if err != nil {
+			return root.ExecutionError(cmd, "Save Error: %s", err)
+		}
+		if resp != nil {
+			return root.ExecutionErrors(cmd, resp.Errors)
 		}
+		return nil
+	}
 
-		if len(savePath) > 0 {
-			// Is the path a directory that exists, or a file to save to?
-
-			// If it is an existing directory? Save under that.
-			var fi fs.FileInfo
-			fi, err = os.Stat(savePath)
-			if err == nil {
-				if fi.IsDir() {
-					filename = filepath.Join(savePath, filename)
-				} else {
-					filename = savePath
-				}
+	blueprint, resp, err := root.Client.GetBlueprintChangeTOML(name, commit)
+	if err != nil {
+		return root.ExecutionError(cmd, "Save Error: %s", err)
+	}
+	if resp != nil && !resp.Status {
+		return root.ExecutionErrors(cmd, resp.Errors)
+	}
+
+	err = saveBlueprint(blueprint, commit, savePath)
+	if err != nil {
+		fmt.Fprintln(os.Stderr, err)
+		return root.ExecutionError(cmd, "")
+	}
+	return nil
+}
+
+// saveBlueprint write the TOML blueprint to a file
+// optionally under a path, or as a new filename
+// and if it is a specific blueprint commit that is appended to the base filename.
+func saveBlueprint(data, commit, path string) error {
+	// Convert the toml blueprint to a struct so we can get the name
+	var bp interface{}
+	err := toml.Unmarshal([]byte(data), &bp)
+	if err != nil {
+		return fmt.Errorf("ERROR: Unmarshal of blueprint failed: %s", err)
+	}
+
+	name, ok := bp.(map[string]interface{})["name"].(string)
+	if !ok {
+		return fmt.Errorf("ERROR: no 'name' in blueprint")
+	}
+
+	// Save to a file in the current directory, replace spaces with - and
+	// remove anything that looks like path separators or path traversal.
+	filename := strings.ReplaceAll(name, " ", "-")
+
+	// If this is a specific blueprint commit, append it to the filename
+	if len(commit) > 0 {
+		filename = filename + "-" + commit
+	}
+	filename = filepath.Base(filename + ".toml")
+	if filename == "/" || filename == "." || filename == ".." {
+		return fmt.Errorf("ERROR: Invalid blueprint filename: %s", name)
+	}
+
+	if len(path) > 0 {
+		// Is the path a directory that exists, or a file to save to?
+
+		// If it is an existing directory? Save under that.
+		var fi fs.FileInfo
+		fi, err = os.Stat(path)
+		if err == nil {
+			if fi.IsDir() {
+				filename = filepath.Join(path, filename)
 			} else {
-				if errors.Is(err, fs.ErrNotExist) {
-					// Does it look like a directory? A directory needs to exist.
-					if savePath[len(savePath)-1] == '/' {
-						fmt.Fprintf(os.Stderr, "ERROR: %s does not exist\n", savePath)
-						rcErr = root.ExecutionError(cmd, "")
-						continue
-					}
-					// Assume it is a file
-					filename = savePath
-				} else {
-					// Some other error
-					fmt.Fprintf(os.Stderr, "ERROR: %s\n", err)
-					rcErr = root.ExecutionError(cmd, "")
-					continue
+				filename = path
```

Comment:
Since all the filename calculation is thrown away when the file is specified, maybe this should be done first.  That way, if a blueprint somehow gets an invalid name with `..` in it for example, a user can still write it out by specifying a filename.
The way it works now, a user would specify a filename but would get an error that the autogenerated filename is invalid, which might make the user think the filename they supplied is invalid (which we should also check for).

Reply:
Ah! Yes, that's a good idea.